datasetUtils/getDatasetDetails.go

```diff
@@ -58,31 +55,23 @@ func GetDatasetDetails(client *http.Client, APIServer string, accessToken string
 
 		datasetDetails, err := fetchDatasetDetails(client, accessToken, myurl)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 
 		for _, datasetId := range datasetList[i:end] {
-			detailsFound := false
-			for _, datasetDetail := range datasetDetails {
-				if datasetDetail.Pid == datasetId {
-					detailsFound = true
-					if ownerGroup == "" || ownerGroup == datasetDetail.OwnerGroup {
-						outputDatasetDetails = append(outputDatasetDetails, datasetDetail)
-						color.Set(color.FgGreen)
-					}
-					log.Printf("%s %9d %v %v\n", datasetId, datasetDetail.Size/1024./1024., datasetDetail.OwnerGroup, datasetDetail.SourceFolder)
-					color.Unset()
-					break
-				}
+			datasetHasIdAndOwnerGroup := func(dataset Dataset) bool {
```

Comment:
do I understand right that this, in addition to the filter [here](https://github.com/paulscherrerinstitute/scicat-cli/pull/115/commits/a5c70ba214dbf1646e528d3c174bb660f4bcc4bf#diff-1aba7554ddcdff9d1dfdba412e079d59438b7ef31ae25f09f3239b7cdd854e9eR48), checks if the ownerGroup is empty or it's equal to dataset.ownerGroup? If so, couldn't one achieve this by modifying the filter? Then (and also if not changing the filter), to find the missing datasets, one could transform datasetDetails into a map and loop over datasetList[i:end]. The ones which are not in the map are the ones missing. This will be O(n) (map creation) + O(m) (datasetList[i:end] loop), while the current implementation is O(~n (indexFunction) *m (loop)) (given the linear search, if I am not mistaken)

Reply:
I think it's also a good idea to add a test for this

Reply:
Yeah, I actually wanted to optimize it but I ended up just simplifying the original code. I'll try to apply your suggestion

Reply:
This should be fixed by the latest commit (should add a test still)

Reply:
Tests were updated, and a new test was added to see if the correct filter is passed

Reply:
thanks!